### ISSUE TITLE
[AIR] `Result` object doesn't work with Ray Client

### ISSUE THREAD

**Issue by Yard1** [ray-project/ray#24396]
### What happened + What you expected to happen

Ran AIR training on the cluster using Ray Client. When I save the results, this is what happens. The files are present on the cluster, but they are not present on the driver, while the experiment analysis is ran on the driver.

```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-17-1e09fdac140e> in <module>
----> 1 result = trainer.fit()

~/ray/python/ray/ml/trainer.py in fit(self)
    328         assert len(result_grid) == 1
    329         try:
--> 330             result = result_grid[0]
    331             if result.error:
    332                 raise result.error

~/ray/python/ray/tune/result_grid.py in __getitem__(self, i)
     86     def __getitem__(self, i) -> Result:
     87         """Returns the i'th result in the grid."""
---> 88         return self._trial_to_result(self._experiment_analysis.trials[i])
     89 
     90     @staticmethod

~/ray/python/ray/tune/result_grid.py in _trial_to_result(self, trial)
    101     def _trial_to_result(self, trial: Trial) -> Result:
    102         if trial.checkpoint.value:
--> 103             checkpoint_dir = TrainableUtil.find_checkpoint_dir(trial.checkpoint.value)
    104             checkpoint = Checkpoint.from_directory(checkpoint_dir)
    105         else:

~/ray/python/ray/tune/utils/trainable.py in find_checkpoint_dir(checkpoint_path)
    116         """
    117         if not os.path.exists(checkpoint_path):
--> 118             raise FileNotFoundError("Path does not exist", checkpoint_path)
    119         if os.path.isdir(checkpoint_path):
    120             checkpoint_dir = checkpoint_path

FileNotFoundError: [Errno Path does not exist] /home/ray/ray_results/HuggingFaceTrainer_2022-05-02_12-12-12/HuggingFaceTrainer_c5a18_00000_0_2022-05-02_12-12-14/checkpoint_000005/
```

### Versions / Dependencies

7cca7782f1648ad77b157d94c0f545cfaea52cc9

### Reproduction script

Run any example we have right now with Ray Client & external cluster.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by xwjiang2010** [comment]
Hmm, can you share your script/setup?

Basically, the file lookup should only happen on head node and the result grid will be streamed back to ray client. Is that aligned with your expectation?

Nonetheless, I think this is a case that we can use some good documentation.

**Comment by Yard1** [comment]
It doesn't look to be happening, though. This is what seems to be happening:
1. TunerInternal finishes fitting on the node, creates a ResultGrid with ExperimentAnalysis pointing to a path on the node
2. ResultGrid is streamed back to the Trainer running on the driver
3. Trainer gets the first item of the ResultGrid - this calls the ExperimentAnalysis contained within. However, now that it's being called on the driver, the path that it's pointing to doesn't exist

It shouldn't happen like that. Instead, we should run the ExperimentAnalysis on the node, and stream the checkpoint back (perhaps lazily, to avoid ser/deser if not needed).

**Comment by xwjiang2010** [comment]
Yeah that makes sense. Let me think that is a good CUJ for this ray client + air story.
Also @richardliaw can I put some color on how invested we are about ray client in the context of AIR?

**Comment by xwjiang2010** [comment]
@Yard1 One implementation is like:
1. convert everything from `experiment_analysis` eagerly in `ResultGrid.__init__`. 
2. overwrite Checkpoint's ser/des behavior: checkpoints are not written to object store by default in this case(that would be a waste!). Instead convert it into Checkpoint(node="xxx", path="yyy") and then serialize it. "xxx" is head node ip in this case. Or we can special case it to be "head".
3. In our documentation, make it clear that if you use ray air with ray client mode, the checkpoint information you get is in this form and the real checkpoint never leaves the cluster.
@Yard1 how does that sound?

**Comment by Yard1** [comment]
I think it's fine if we just use object ref checkpoints and don't make any special provisions for client mode aside from that. We can pass those around without issues, they are very small.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
